addons/sale_timesheet/models/sale_order.py

```diff
@@ -51,10 +54,14 @@ def _compute_project_ids(self):
     def _action_confirm(self):
         """ On SO confirmation, some lines should generate a task or a project. """
         result = super(SaleOrder, self)._action_confirm()
-        self.mapped('order_line').sudo().with_context(
-            default_company_id=self.company_id.id,
-            force_company=self.company_id.id,
-        )._timesheet_service_generation()
+        for company, lines in groupby(
+            self.mapped('order_line').sudo(),
+            key=operator.itemgetter('company_id'),
+        ):
+            self.env['sale.order.line'].concat(*lines).sudo().with_context(
```

Comment:
seems like you're trying to make your code difficult to read on purpose :p
Why not something much more readable, like
```py
companies = self.mapped('company_id')
for company in companies:
    company_ctx = {'default_company_id': company.id, 'force_company': company.id}
    orders = self.filtered(lambda so: so.company_id == company).with_context(**company_ctx)
    orders.mapped('order_line').sudo()._timesheet_service_generation()
```

- easier to read
- no new imports
- no `groupby` or `itemgetter` which, sure, work - but you always need to check their args and exact workings and I just don't like those if you can do without